### PR TITLE
Add IterDomainBuilder

### PR DESCRIPTION
**Pull request by csarofeen**
Refactor construction of IterDomains, two minor (unrelated) cleanups included.
IterDomainBuilder's goal isn't really to be less verbose, just more explicit about which properties we really care about setting without having to duplicate default values everywhere when constructing IterDomains. Some minor cleanup here in there included as well as no one should really be manually constructing IterDomains outside of our internals.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -193,8 +193,11 @@ void initNvFuserPythonBindings(PyObject* module) {
           py::arg("dtype") = torch::jit::fuser::cuda::DataType::Float,
           py::return_value_policy::reference)
       .def(
+          // TODO: Should the inernals of this function live more explicitly in
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:196]: I don't have a known way to test this. @kevinstephano can you check this change and make sure it works please?

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 2]
```diff
@@ -641,6 +641,52 @@ class TORCH_CUDA_CU_API LoadStoreOp : public Expr {
...
+// all the default values. Intended to be used with IterDomain::IterDomain(IrBuilderPasskey IterDomainBuildArgs)
+class TORCH_CUDA_CU_API IterDomainBuilder{
+public:
+ // Match legacy constructor
+ IterDomainBuilder(Val* _start, Val* _extent);
+
+ // Grab all the parameters from id to set the IterDomainBuilder
+ IterDomainBuilder(const IterDomain* id);
+
+ // Resets defaults for rfactor, is padded dim, padded to size, and is mma
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: Isn't the parallel type also considered a scheduling parameter?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 3]
```diff
@@ -68,9 +68,9 @@ void validateDoubleBufferedTensor(const TensorView* tv) {
   auto def = tv->definition();
   TORCH_INTERNAL_ASSERT(
       def->isA<UnaryOp>() &&
-              def->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Set ||
-          // Load store op should generally support double buffering.
-          def->isA<LoadStoreOp>(),
+          (def->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Set ||
+           // Load store op should generally support double buffering.
+           def->isA<LoadStoreOp>()),
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp]: I don't think this is right as `def` can be `LoadStoreOp`.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp]: Already fixed #1732

File: torch/csrc/jit/codegen/cuda/transform_rfactor.cpp [hunk 4]
```diff
@@ -117,24 +117,26 @@ class ReplayRFactor : public ReplayTransformations {
...
-        s->container(),
-        IrBuilder::create<Int>(s->container(), 0),
-        s->innerSplit() ? s->factor() : remainder->as<Int>(),
-        ParallelType::Serial,
-        rfactor_axes_.count(s->inner()) ? IterType::Reduction
-                                        : IterType::Iteration,
-        static_rfactor_outputs);
+    IterDomain* idi =
+        IterDomainBuilder(
+            IrBuilder::create<Int>(s->container(), 0),
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_rfactor.cpp]: `s->container()->zeroVal()`

File: torch/csrc/jit/codegen/cuda/transform_rfactor.cpp [hunk 5]
```diff
@@ -179,14 +181,15 @@ class ReplayRFactor : public ReplayTransformations {
...
-        m->container(),
-        IrBuilder::create<Int>(m->container(), 0),
-        merged_id_size->as<Int>(),
-        ParallelType::Serial,
-        rfactor_axes_.count(m->out()) ? IterType::Reduction
-                                      : IterType::Iteration,
-        static_rfactor_ids_.count(m->out()));
+    IterDomain* merged_id =
+        IterDomainBuilder(
+            IrBuilder::create<Int>(m->container(), 0),
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_rfactor.cpp]: ditto

**Review by naoyam — APPROVED** [review]
Nice cleanup!

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: Yeah, good point.

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp]: Reverting.

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp (hunk 1 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:196]: The examples work with the change.

**Review by kevinstephano — APPROVED** [review]
LGTM

**Pull request merged**